src/annotator/integrations/html-side-by-side.js

```diff
@@ -0,0 +1,199 @@
+    leftVotes += 1;
+    leftMarginVotes.set(rect.left, leftVotes);
+
+    let rightVotes = rightMarginVotes.get(rect.right) ?? 0;
+    rightVotes += 1;
+    rightMarginVotes.set(rect.right, rightVotes);
+  });
+
+  // Find the margin values with the most votes.
+  if (leftMarginVotes.size === 0 || rightMarginVotes.size === 0) {
+    return null;
+  }
+
+  const leftMargin = [...leftMarginVotes.entries()].sort((a, b) => b[1] - a[1]);
+  const rightMargin = [...rightMarginVotes.entries()].sort(
+    (a, b) => b[1] - a[1]
+  );
+
+  const [leftPos] = leftMargin[0];
+  const [rightPos] = rightMargin[0];
+
+  return { left: leftPos, right: rightPos };
+}
+
+/** @type {Range} */
+let textRectRange;
+
+/**
+ * Return the viewport-relative rect occupied by part of a text node.
+ *
+ * @param {Text} text
+ * @param {number} start
+ * @param {number} end
+ */
+function textRect(text, start = 0, end = text.data.length) {
+  if (!textRectRange) {
+    // Allocate a range only on the first call to avoid the overhead of
+    // constructing and maintaining a large number of live ranges.
+    textRectRange = document.createRange();
+  }
+  textRectRange.setStart(text, start);
+  textRectRange.setEnd(text, end);
+  return textRectRange.getBoundingClientRect();
+}
+
+/**
+ * Yield all the text node descendants of `root` that intersect `rect`.
+ *
+ * @param {Element} root
+ * @param {DOMRect} rect
+ * @return {Generator<Text>}
+ */
+function* textNodesInRect(root, rect) {
+  /** @type {Node|null} */
+  let node = root.firstChild;
+  while (node) {
+    if (node.nodeType === Node.ELEMENT_NODE) {
+      const element = /** @type {Element} */ (node);
+      const elementRect = element.getBoundingClientRect();
+
+      // Skip over subtrees which are entirely outside the viewport or hidden.
+      if (rectIntersects(elementRect, rect)) {
+        yield* textNodesInRect(element, rect);
+      }
+    } else if (node.nodeType === Node.TEXT_NODE) {
+      const text = /** @type {Text} */ (node);
+
+      // Skip over text nodes which are entirely outside the viewport or empty.
+      if (rectIntersects(textRect(text), rect)) {
+        yield text;
+      }
+    }
+    node = node.nextSibling;
+  }
+}
+
+/**
+ * Find content within an element to use as an anchor when scrolling.
+ *
+ * @param {Element} scrollRoot
+ * @return {Range|null} - Range to use as an anchor or `null` if a suitable
+ *   range could not be found
+ */
+function getScrollAnchor(scrollRoot) {
+  // Range representing the content whose position within the viewport we will
+  // try to maintain after running the callback.
+  let anchorRange = /** @type {Range|null} */ (null);
+
+  const viewport = intersectRects(
+    scrollRoot.getBoundingClientRect(),
+    new DOMRect(0, 0, window.innerWidth, window.innerHeight)
+  );
+  if (viewport.width < 0 || viewport.height < 0) {
+    // Element being scrolled is outside the viewport
+    return null;
+  }
```

Comment:
In our applied use, the `scrollRoot` is always the `document` element, correct? So this case would not typically happen in our reality? (It's good to handle it no matter what, of course; I'm just curious if this could ever really happen on the `document` element as `scrollRoot`).

Reply:
Yes, you're correct. The `scrollRoot` is primarily useful in testing.